### ISSUE TITLE
jack cannot use realtime scheduling

### ISSUE THREAD

**Issue by VoidDuck** [void-linux/void-packages#20051]
Hello,

I tried many different settings but I definitely cannot get jack to run with realtime priority on Void, while it works fine on other distributions.

I checked everything described on https://jackaudio.org/faq/linux_rt_config.html and https://github.com/jackaudio/jackaudio.github.com/wiki/Cgroups but it still does not work.

Here is what I get when running `jackd -R -d alsa`:

> JACK server starting in realtime mode with priority 10
> self-connect-mode is "Don't restrict self connect requests"
> Cannot lock down 82280346 byte memory area (Cannot allocate memory)
> audio_reservation_init
> Acquire audio card Audio0
> creating alsa driver ... hw:0|hw:0|1024|2|48000|0|0|nomon|swmeter|-|32bit
> configuring for 48000Hz, period = 1024 frames (21.3 ms), buffer = 2 periods
> ALSA: final selected sample format for capture: 32bit integer little-endian
> ALSA: use 2 periods for capture
> ALSA: final selected sample format for playback: 32bit integer little-endian
> ALSA: use 2 periods for playback
> Cannot use real-time scheduling (RR/10) (1: Operation not permitted)
> AcquireSelfRealTime error

### System
* xuname:  
  *Void 5.4.24_1 x86_64 GenuineIntel uptodate rF*
* package:  
  *jack-1.9.14_1*

**Comment by VoidDuck** [comment]
For your information, I just tried the same in a root session and it works perfectly there, so the problem definitely is on the user permissions side.

**Comment by daniel-eys** [comment]
Your user is definitely added to the audio group?
```sh
groups | grep audio
```

**Comment by VoidDuck** [comment]
Yes.

**Comment by daniel-eys** [comment]
What's the output of `ulimit -l` for your user?
It should output `unlimited`.

**Comment by jrigg** [comment]
What happens if you do
`export JACK_NO_AUDIO_RESERVATION=TRUE`
before starting jackd? That disables jackdbus and runs plain jackd.

I have the above set permanently in .bashrc and jackd has always worked properly with realtime priority on my Void system.

**Comment by VoidDuck** [comment]
> What's the output of `ulimit -l` for your user?
> It should output `unlimited`.
I get `64`...


> What happens if you do
> `export JACK_NO_AUDIO_RESERVATION=TRUE`
> before starting jackd?
Nothing special. I get the same output as usual.

**Comment by daniel-eys** [comment]
> I get `64`...

This is the issue here.
The jack package installs `/etc/security/limits.d/jack.conf` which permits the audio group to access higher resource limits:
`@audio	-	memlock	unlimited`.

After adding your user to the audio group, have you logged out and in again for the change to take effect?

**Comment by VoidDuck** [comment]
Definitely, my user has been in the audio group since I installed the system.

In `/etc/group` I have `audio:x:12:myuser`

In `/etc/security/limits.d/jack.conf` I have
`@audio  -       rtprio  99`
`@audio  -       memlock unlimited`

**Comment by daniel-eys** [comment]
And you did a re-login after adding your user to audio? (There needs to be a new session such that pam picks up and sets all limits)
Have you manually changed `memlock` within `/etc/security/*` perhaps somewhere?
Memlock should be set to unlimited for all audio group members.

**Comment by VoidDuck** [comment]
Don't worry, I have been logging in and out dozens of times in my attempts to get jack working properly...

The only other memlock-related setting I can find in `/etc/security` is `limits.d/10-gcr-memlock.conf`:
`# Increate max locked memory to 1024KB for the users group.`
`@users - memlock 1024`

**Comment by daniel-eys** [comment]
Right, this restricts the group `users` to limit memlock to 1024KB.
But you said your memlock limit is still at 64. (which is definitely the issue here)
I'm not really sure why jack.conf is not properly picked up and setting memlock to unlimited. :/
Are you able to manually `ulimit -l unlimited`?

**Comment by VoidDuck** [comment]
No, it does not work either.
`bash: ulimit: max locked memory: cannot modify limit: Operation not permitted`

**Comment by daniel-eys** [comment]
I'm running a bit out of ideas here.
Fresh void iso, logging in as `anon` user (anon is in the audio group)

```sh
[anon@void-live ~]$ ulimit -l
64
```

After installing `jack` and re-login:

```sh
[anon@void-live ~]$ ulimit -l
unlimited
```

Are there maybe any `ulimit`s set in your global bash-setup?
Any modifications within `/etc/pam.d/`?

**Comment by VoidDuck** [comment]
I could not find anything related in /etc/pam.d/ nor in my bash settings.

However I tried logging in as user in text mode and there `ulimit -l` gave me `unlimited` indeed. Back to the graphical interface (I use Xfce and LightDM) I get `64` again... may the problem come from LightDM?

**Comment by VoidDuck** [comment]
That's it! If I start Xfce from command line without using LightDM I keep the unlimited memlock. Unless someone has an idea about solving that problem I'm just getting rid of LightDM. Thank you @daniel-eys for your help.

**Comment by daniel-eys** [comment]
Interesting, so lightdm caused this.
Looking at `/etc/pam.d/lightdm`, lightdm seems to delegate pam to `pam_elogind.so` (?), but I don't really know where/when/how elogind sets limits for its sessions.
One workaround would be to explicitly load `pam_limits` (the pam module responsible for setting resource limits) by adding

`session required pam_limits.so`

to `/etc/pam.d/lightdm`, like so:
```
...
# Setup session
session   required pam_unix.so
session   required pam_limits.so
-session   optional pam_elogind.so
-session optional pam_ck_connector.so nox11
```

So with this added, you could actually continue using lightdm. @VoidDuck

**Comment by VoidDuck** [comment]
It works. Thank you very much!

In this case I think that line should be added to the default `/etc/pam.d/lightdm` shipped in Void.
Are you able to do this or should I open another issue about it?

**Comment by daniel-eys** [comment]
You're welcome, glad to hear that it works :)

My issue here is that I'm not really sure whether this is "the proper" fix to be included in the lightdm package. This was more like my first thought how it might work and it actually did.
For example, there is also polkit which handles `pam_limits`, maybe it's elogind's job to set session limits, `gdm` and `slim` add `session include system-local-login`, which (after some redirections) pulls in `pam_limits` via `system-auth`, I'm still a bit confused here.
